### PR TITLE
[REF] sale_coupon: create coupon module

### PR DESCRIPTION
**Pull request by caburj**
This refactoring extracts parts of `sale_coupon` module to create a more
generic module called `coupon`. This new module can be extended in
different sales modules to implement functionalities specific to modules
that extend it.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/coupon/wizard/coupon_generate.py [hunk 1]
```diff
@@ -5,7 +5,7 @@
 
 
 class SaleCouponGenerate(models.TransientModel):
```
Comment (mreficent) [inline addons/coupon/wizard/coupon_generate.py]:
```suggestion
class CouponGenerate(models.TransientModel):
```

File: addons/coupon/wizard/coupon_generate.py [hunk 2]
```diff
@@ -5,7 +5,7 @@
...
-    _name = 'sale.coupon.generate'
+    _name = 'coupon.coupon.generate'
     _description = 'Generate Sales Coupon'
```
Comment (mreficent) [inline addons/coupon/wizard/coupon_generate.py]:
```suggestion
    _description = 'Generate Coupon'
```

File: addons/sale_coupon/models/sale_coupon.py [hunk 3]
```diff
@@ -7,47 +7,12 @@
 
 
 class SaleCoupon(models.Model):
```
Comment (mreficent) [inline addons/sale_coupon/models/sale_coupon.py]:
```suggestion
class Coupon(models.Model):
```

File: addons/sale_coupon/models/sale_coupon_program.py [hunk 4]
```diff
@@ -7,58 +7,9 @@
 
 
 class SaleCouponProgram(models.Model):
```
Comment (mreficent) [inline addons/sale_coupon/models/sale_coupon_program.py]:
```suggestion
class CouponProgram(models.Model):
```

File: addons/coupon/wizard/coupon_generate.py (hunk 1 above)
Comment (caburj) [reply addons/coupon/wizard/coupon_generate.py]: @mreficent Thanks for the suggestions man. I'll take them into consideration.

**Comment by caburj** [comment]
Closing this pull request in favor of: https://github.com/odoo/odoo/pull/38382

**Pull request closed without merging**